File: src/stars.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { initial, wallBonus } from "./game.js";
const unit = 22.4 / 576;
test("all ten stars use complete original terrain cells, including edges", () => {
  for (const group of [0, 1])
    for (let index = 0; index < 5; index++)
      for (const offset of [0, 16, 31]) {
        const s = initial(),
          start = group === 0 ? 384 : 608;
        Object.assign(s.ball, {
          owner: -1,
          lastTouch: group * 9 + 7,
          x: (576 - start - index * 32 - offset) * unit,
          z: (group === 0 ? -288 : 288) * unit,
          h: 4,
        });
        wallBonus(s);
        assert.equal(s.stars[group], 1 << index);
        assert.equal(s.score[group], 2);
      }
});
test("star bank excludes adjacent cells and interior balls; throwing carrier is permitted", () => {
  for (const group of [0, 1]) {
    const start = group === 0 ? 384 : 608;
    for (const [y, z] of [
      [start - 1, group === 0 ? -288 : 288],
      [start + 160, group === 0 ? -288 : 288],
      [start + 16, group === 0 ? -287 : 287],
    ]) {
      const s = initial();
      Object.assign(s.ball, {
        owner: -1,
        lastTouch: group * 9 + 7,
        x: (576 - y) * unit,
        z: z * unit,
      });
      wallBonus(s);
      assert.deepEqual(s.stars, [0, 0]);
    }
  }
  const s = initial();
  Object.assign(s.ball, { owner: 7, lastTouch: 7, x: 176 * unit, z: -11.2 });
  wallBonus(s);
  assert.equal(s.score[0], 0);
  s.players[7].action = 3;
  wallBonus(s);
  assert.equal(s.score[0], 2);
});

File: server/stars_test.go
package main

import "testing"

func TestOriginalStarCells(t *testing.T) {
	for group := 0; group < 2; group++ {
		start, z := 384, -288
		if group == 1 {
			start, z = 608, 288
		}
		for index := 0; index < 5; index++ {
			for _, offset := range []int{0, 16, 31} {
				s := initial()
				s.Ball = Ball{Owner: -1, LastTouch: group*9 + 7, X: float64(576-start-index*32-offset) * terrainUnit, Z: float64(z) * terrainUnit, H: 4}
				s.wallBonus()
				if s.Stars[group] != 1<<index || s.Score[group] != 2 {
					t.Fatal("star cell", group, index, offset)
				}
			}
		}
		for _, y := range []int{start - 1, start + 160} {
			s := initial()
			s.Ball = Ball{Owner: -1, LastTouch: group*9 + 7, X: float64(576-y) * terrainUnit, Z: float64(z) * terrainUnit}
			s.wallBonus()
			if s.Score[group] != 0 {
				t.Fatal("outside bank scored")
			}
		}
	}
}
